File: LinkedList/js/DoublyLinkedList.js
"use strict";


class Node {

    constructor(element) {
        this._element = element;
        this._next = null;
        this._previous = null;
    }

    get next() {
        return this._next;
    }

    set next(node) {
        this._next = node;
        if (node != null) {
            node._previous = this;
        }
    }

    get previous() {
        return this._previous;
    }

    set previous(node) {
        this._previous = node;
        if (node != null) {
            node._next = this;
        }
    }

    get element() {
        return this._element;
    }

    set element(element) {
        this._element = element;
    }

    toString() {
        let n = this._next != null ? this._next.element : null;
        let p = this._previous != null ? this._previous.element : null;
        return `Node(element=${this._element}, next=${n}, previous=${p})`;
    }
}

class DoublyLinkedList {

    constructor() {
        this._head = null;
        this._tail = null;
        this._size = 0;
    }

    get size() {
        return this._size;
    }

    get head() {
        return this._head;
    }

    set head(node) {
        if (this._head != null) {
            this._head.previous = node;
        }
        this._head = node;
    }

    get tail() {
        return this._tail;
    }

    set tail(node) {
        if (this._tail != null) {
            this._tail.next = node;
        }
        this._tail = node;
    }

    get isEmpty() {
        return this._size < 1;
    }

    append(node) {
        this.tail = node;
        if (this._head === null) {
            this.head = node;
        }
        this._size++;
    }

    prepend(node) {
        this.head = node;
        if (this._tail === null) {
            this.tail = node;
        }
        this._size++;
    }

    remove(node) {
        let next = node.next;
        let previous = node.previous;
        if (next != null) {
            next.previous = previous;
        } else if (previous != null) {
            previous.next = null;
        }
        node.next = null;
        node.previous = null;
        this._size--;
    }
}

const lst = new DoublyLinkedList();
lst.append(new Node(1));
lst.append(new Node(2));
lst.append(new Node(3));
lst.append(new Node(5));
lst.prepend(new Node(0));
console.log(lst.size);
console.log(lst.head);
console.log(lst.tail);

File: LinkedList/dart/bin/doubly_linked_list.dart
void main() {
  DoublyLinkedList<int> lst = new DoublyLinkedList();
  lst.append(2);
  lst.append(4);
  lst.append(2);
  lst.prepend(46);
  print(lst.size);
  print(lst.head);
  print(lst.tail);

  print("\nLooping...\n");
  for (Node<int> n in lst) {
    print(n);
  }
}

class Node<E> {
  E _element;
  Node<E>? _next;
  Node<E>? _previous;

  Node(this._element, [Node<E>? next, Node<E>? previous]) {
    this._next = next;
    this._previous = previous;
  }

  E get element => _element;

  void set element(E element) {
    this._element = element;
  }

  Node<E>? get next => _next;

  void set next(Node<E>? node) {
    this._next = node;
    node?._previous = this;
  }

  Node<E>? get previous => _previous;

  void set previous(Node<E>? node) {
    this._previous = node;
    node?._next = this;
  }

  @override
  String toString() {
    return "Node(element=$_element, next=${_next?._element}, previous=${_previous?._element})";
  }
}

class DoublyLinkedListIterator<E> implements Iterator<Node<E>> {
  Node<E>? _next;

  DoublyLinkedListIterator(this._next);

  Node<E> get current {
    Node<E> n = _next!;
    this._next = n.next;
    return n;
  }

  @override
  bool moveNext() {
    return _next != null;
  }
}

class DoublyLinkedList<E> extends Iterable<Node<E>> {
  Node<E>? _head;
  Node<E>? _tail;
  int _size = 0;

  DoublyLinkedList([Node<E>? head, Node<E>? tail]) {
    this._head = head;
    this._tail = tail;
  }

  Node<E>? get head => _head;

  void set head(Node<E>? node) {
    _head?.previous = node;
    this._head = node;
  }

  Node<E>? get tail => _tail;

  void set tail(Node<E>? node) {
    _tail?.next = node;
    this._tail = node;
  }

  int get size => _size;

  bool get isEmpty => _size < 1;

  void append(E element) {
    Node<E> node = Node(element);
    this.tail = node;
    if (_head == null) {
      this.head = node;
    }
    _size++;
  }

  void prepend(E element) {
    Node<E> node = Node(element);
    this.head = node;
    if (_tail == null) {
      this.tail = node;
    }
    _size++;
  }

  void remove(Node<E> node) {
    Node<E>? next = node.next;
    Node<E>? previous = node.previous;

    if (next != null) {
      next.previous = previous;
    } else if (previous != null) {
      previous.next = null;
    }

    node.next = null;
    node.previous = null;
    _size--;
  }

  @override
  Iterator<Node<E>> get iterator => DoublyLinkedListIterator(_head);
}
